### ISSUE TITLE
[v5.0.1] Getting locale from session behavior has been changed?

### ISSUE THREAD

**Issue by fukayatsu** [r18n/sinatra-r18n#3]
> https://github.com/r18n/sinatra-r18n#how-to

Getting locale from session behavior seems changed on v5.0.1.
Is this an intentional change?

### Steps to reproduce

**Update: Put debug.rb into an empty folder**
**Update 2: Add `--options /dev/null` to skip rspec options file like `.rspec` or `.rspec-local`**

1. put `debug.rb` ~~into an empty folder~~
2.  `$ ruby debug.rb --options /dev/null`

### Expected behavior

Exit with no failures.

### Actual behavior

3 examples, 1 failure

### Other Information

```
$  ruby -v
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
```

### `debug.rb`: 

**Update: Add sinatra, pry-byebug, simplecov to `debug.rb`**

```ruby
#!/usr/bin/env ruby


### Setup ###

require 'bundler/inline'

gemfile do
  gem 'sinatra', '2.1.0'

  # gem 'sinatra-r18n', '4.0.0' #=> OK: 3 examples, 0 failures
  gem 'sinatra-r18n', '5.0.1' #=> NG: 3 examples, 1 failures

  gem 'rack-test', '~> 1.1'
  gem 'rspec', '~> 3.10'
  gem 'pry-byebug', '~> 3.9'
  gem 'simplecov', '~> 0.21.0'
end

require 'rspec/autorun'



### Dummy App ###

require 'sinatra/base'
require 'sinatra/r18n'
class App < Sinatra::Base
  register Sinatra::R18n
  set :root, __dir__
  R18n::I18n.default = 'ru'

  # Need to define before `register Sinatra::R18n` on sinatra-r18n v5.0.1?
  before do
    session[:locale] = params[:lang] if params[:lang]
  end

  get '/show_locale' do
    R18n.get.locales.first.code
  end
end


### Test ###

module RSpecMixinExample
  include Rack::Test::Methods

  def app
    App.new
  end
end

RSpec.configure { |c| c.include RSpecMixinExample }

RSpec.describe Sinatra::R18n do
  after do
    ::R18n.thread_set nil
  end

  describe 'locale from session[:locale]' do
    context 'no param' do
      it 'Show default locale' do
        get '/show_locale'
        expect(last_response.body).to eq('ru')
      end
    end

    context 'lang=ru' do
      it 'Show locale=ru' do
        get '/show_locale?lang=ru'
        expect(last_response.body).to eq('ru')
      end
    end

    context 'lang=en' do
      # This example fails on v5.0.1
      # expected: "en"
      #      got: "ru"
      it 'Show locale=en' do
        get '/show_locale?lang=en'
        expect(last_response.body).to eq('en')
      end
    end
  end
end
```

**Comment by AlexWayfer** [comment]
Thank you for detailed report! I'd be glad to help you, but I've faced some problems with reproduction:

1.  I've added gems to make this `debug.rb` work in the repo folder:
    ```ruby
    gem 'pry-byebug', '~> 3.9'
    gem 'simplecov', '~> 0.21.0'
    ```
2. I have a different output:

```
> bundle exec ruby debug.rb
/home/alex/Projects/ruby/sinatra-r18n/spec/spec_helper.rb:22: warning: method redefined; discarding old app
debug.rb:47: warning: previous definition of app was here

Randomized with seed 21929

Sinatra::R18n
  locale from session[:locale]
    lang=en
      Show locale=en (FAILED - 1)
    no param
      Show default locale (FAILED - 2)
    lang=ru
      Show locale=ru (FAILED - 3)

Failures:

  1) Sinatra::R18n locale from session[:locale] lang=en Show locale=en
     Failure/Error: expect(last_response.body).to eq('en')
     
       expected: "en"
            got: "<!DOCTYPE html>\n<html>\n<head>\n  <style type=\"text/css\">\n  body { text-align:center;font-family...27;&#x2F;show_locale&#x27; do\n  &quot;Hello World&quot;\nend\n</pre>\n  </div>\n</body>\n</html>\n"
     
       (compared using ==)
     
       Diff:
       @@ -1,21 +1,41 @@
       -en
       +<!DOCTYPE html>
       +<html>
       +<head>
       +  <style type="text/css">
       +  body { text-align:center;font-family:helvetica,arial;font-size:22px;
       +    color:#888;margin:20px}
       +  #c {margin:0 auto;width:500px;text-align:left}
       +  </style>
       +</head>
       +<body>
       +  <h2>Sinatra doesn’t know this ditty.</h2>
       +  <img src='http://example.org/__sinatra__/404.png'>
       +  <div id="c">
       +    Try this:
       +    <pre>get &#x27;&#x2F;show_locale&#x27; do
       +  &quot;Hello World&quot;
       +end
       +</pre>
       +  </div>
       +</body>
       +</html>
       
     # debug.rb:80:in `block (4 levels) in <main>'

  2) Sinatra::R18n locale from session[:locale] no param Show default locale
     Failure/Error: expect(last_response.body).to eq('ru')
     
       expected: "ru"
            got: "<!DOCTYPE html>\n<html>\n<head>\n  <style type=\"text/css\">\n  body { text-align:center;font-family...27;&#x2F;show_locale&#x27; do\n  &quot;Hello World&quot;\nend\n</pre>\n  </div>\n</body>\n</html>\n"
     
       (compared using ==)
     
       Diff:
       @@ -1,21 +1,41 @@
       -ru
       +<!DOCTYPE html>
       +<html>
       +<head>
       +  <style type="text/css">
       +  body { text-align:center;font-family:helvetica,arial;font-size:22px;
       +    color:#888;margin:20px}
       +  #c {margin:0 auto;width:500px;text-align:left}
       +  </style>
       +</head>
       +<body>
       +  <h2>Sinatra doesn’t know this ditty.</h2>
       +  <img src='http://example.org/__sinatra__/404.png'>
       +  <div id="c">
       +    Try this:
       +    <pre>get &#x27;&#x2F;show_locale&#x27; do
       +  &quot;Hello World&quot;
       +end
       +</pre>
       +  </div>
       +</body>
       +</html>
       
     # debug.rb:63:in `block (4 levels) in <main>'

  3) Sinatra::R18n locale from session[:locale] lang=ru Show locale=ru
     Failure/Error: expect(last_response.body).to eq('ru')
     
       expected: "ru"
            got: "<!DOCTYPE html>\n<html>\n<head>\n  <style type=\"text/css\">\n  body { text-align:center;font-family...27;&#x2F;show_locale&#x27; do\n  &quot;Hello World&quot;\nend\n</pre>\n  </div>\n</body>\n</html>\n"
     
       (compared using ==)
     
       Diff:
       @@ -1,21 +1,41 @@
       -ru
       +<!DOCTYPE html>
       +<html>
       +<head>
       +  <style type="text/css">
       +  body { text-align:center;font-family:helvetica,arial;font-size:22px;
       +    color:#888;margin:20px}
       +  #c {margin:0 auto;width:500px;text-align:left}
       +  </style>
       +</head>
       +<body>
       +  <h2>Sinatra doesn’t know this ditty.</h2>
       +  <img src='http://example.org/__sinatra__/404.png'>
       +  <div id="c">
       +    Try this:
       +    <pre>get &#x27;&#x2F;show_locale&#x27; do
       +  &quot;Hello World&quot;
       +end
       +</pre>
       +  </div>
       +</body>
       +</html>
       
     # debug.rb:70:in `block (4 levels) in <main>'

Finished in 0.14538 seconds (files took 0.12248 seconds to load)
3 examples, 3 failures

Failed examples:

rspec debug.rb:78 # Sinatra::R18n locale from session[:locale] lang=en Show locale=en
rspec debug.rb:61 # Sinatra::R18n locale from session[:locale] no param Show default locale
rspec debug.rb:68 # Sinatra::R18n locale from session[:locale] lang=ru Show locale=ru

Randomized with seed 21929

Coverage report generated for RSpec to /home/alex/Projects/ruby/sinatra-r18n/coverage. 10 / 19 LOC (52.63%) covered.
Stopped processing SimpleCov as a previous error not related to SimpleCov has been detected
```

**Comment by AlexWayfer** [comment]
BTW, here is my `bundle list`:

```
> bundle list
Gems included by the bundle:
  * alt_memery (2.1.0)
  * ast (2.4.2)
  * byebug (11.1.3)
  * codecov (0.5.1)
  * coderay (1.1.3)
  * diff-lcs (1.4.4)
  * docile (1.3.5)
  * faraday (1.3.0)
  * faraday-net_http (1.0.1)
  * faraday_middleware (1.0.0)
  * gem_toys (0.7.1)
  * method_source (1.0.0)
  * module_methods (0.1.0)
  * multipart-post (2.1.1)
  * mustermann (1.1.1)
  * parallel (1.20.1)
  * parser (3.0.0.0)
  * pry (0.13.1)
  * pry-byebug (3.9.0)
  * r18n-core (5.0.1 275708f)
  * rack (2.2.3)
  * rack-protection (2.1.0)
  * rack-test (1.1.0)
  * rainbow (3.0.0)
  * regexp_parser (2.1.1)
  * rexml (3.2.4)
  * rspec (3.10.0)
  * rspec-core (3.10.1)
  * rspec-expectations (3.10.1)
  * rspec-mocks (3.10.2)
  * rspec-support (3.10.2)
  * rubocop (1.11.0)
  * rubocop-ast (1.4.1)
  * rubocop-performance (1.10.1)
  * rubocop-rake (0.5.1)
  * ruby-progressbar (1.11.0)
  * ruby2_keywords (0.0.4)
  * simplecov (0.21.2)
  * simplecov-html (0.12.3)
  * simplecov_json_formatter (0.1.2)
  * sinatra (2.1.0)
  * sinatra-r18n (5.0.1)
  * tilt (2.0.10)
  * toys (0.11.4)
  * toys-core (0.11.4)
  * unicode-display_width (2.0.0)
Use `bundle info` to print more detailed information about a gem
```

Maybe we have different versions of `sinatra` of something else?

**Comment by fukayatsu** [comment]
@AlexWayfer Thank you for your reply.

> 1. I've added gems to make this debug.rb work in the repo folder
> 2. I have a different output:

I'm sorry for that. I've been put `debug.rb` into an empty folder. 
I updated the description.

I confirm the same output as yours when I put `debug.rb` in the repo folder.
In the repo folder, `.rspec` requires `spec_helper`, which causes the problem.

**Comment by fukayatsu** [comment]
I found adding `--options /dev/null` is enough to skip loading `.rspec`.
I updated the description again.

**Comment by AlexWayfer** [comment]
@fukayatsu thank you for all your information! It helped very much.

It was not easy to investigate, especially due to mono-repo splitting, but I've done it.

1. Your code-comment in the example app is correct: please, register R18n after your hooks, because `Sinatra::R18n` depends on data from `params` and `session`.
2. The reason of changes is https://github.com/r18n/r18n/pull/236/files#diff-db454afb38d4e7f749d69d5d647c2aed63ee76ec153dd31608b03e28dcee0954
    I didn't think that it'll be so breaking because formally `R18n#thread_set` receives either arg or block, but, I'm almost sure that with the block there was lazy evaluation.

I'm leaving this issue open until I'll write this info in the README and also I didn't find specs for examples like your one.

I hope you're OK with the `register Sinatra::R18n` moving, also there was a major release so formally we're allowed to make breaking changes (I'm glad that it works at least in some way 😅 ).

**Comment by AlexWayfer** [comment]
Also, `#thread_set` with block was removed separately: https://github.com/r18n/r18n-core/commit/593cbe667d0759391dea6d14c0b591544ded4609

I've got this as unnecessary legacy, but it appears to working different.

@fukayatsu, as I understand, you're familiar with Sinatra and its eco-system, so what do you think? Should we return this lazy evaluation or it's OK to left it simpler and require the order of R18n-related hooks and R18n registration.

**Comment by fukayatsu** [comment]
@AlexWayfer Thank you for your investigation!

> I'm leaving this issue open until I'll write this info in the README and also I didn't find specs for examples like your one.

That's nice!
And it would be very helpful if you could write that info in the ChangeLog too.

> I hope you're OK with the register Sinatra::R18n moving, also there was a major release so formally we're allowed to make breaking changes (I'm glad that it works at least in some way 😅 ).

👍 

> @fukayatsu, as I understand, you're familiar with Sinatra and its eco-system, so what do you think? Should we return this lazy evaluation or it's OK to left it simpler and require the order of R18n-related hooks and R18n registration.

I'm not so familiar with Sinatra's eco-system, but I like Sinatra's simplicity, so I think it's OK to cut off that lazy evaluation on sinatra-r18n v5.0.

**Comment by AlexWayfer** [comment]
Done in #4, released as 5.0.2. Thank you for the report and for help!
